File: proposed-projects/round-trip-experiment/reverse/20260909T154949Z_461ef9c5/20260910T092120Z_46cf829b/ported_implementation/src/utils/errors/input-parse-error.ts
import type { ValidInput } from '../../valid-input.js';
import { buildErrorPosition } from './build-error-position.js';
import { getInputAsString } from './get-input-as-string.js';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

export class InputParseError extends Error {
  private mostRecentInput: ValidInput;
  private pos: number;
  private previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    const previous = getInputAsString(previousInput);
    super([
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(
        `${previous}${getInputAsString(mostRecentInput)}`,
        pos + previous.length,
      ),
    ].join('\n'));
    this.name = 'InputParseError';
    this.mostRecentInput = mostRecentInput;
    this.pos = pos;
    this.previousInput = previousInput;
  }

  get src(): string {
    return `${getInputAsString(this.previousInput)}${getInputAsString(this.mostRecentInput)}`;
  }

  get errorForMostRecentInput(): string {
    return [
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(getInputAsString(this.mostRecentInput), this.pos),
    ].join('\n');
  }
}

File: packages/gbnf-experiment/data/20260909T154949Z_461ef9c5/ported_implementation/gbnf/utils/errors/input_parse_error.py
"""Port of ``src/utils/errors/input-parse-error.ts``."""
from __future__ import annotations

from ...valid_input import ValidInput
from .build_error_position import build_error_position
from .get_input_as_string import get_input_as_string

INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:'


class InputParseError(Exception):
    def __init__(self, most_recent_input: ValidInput, pos: int, previous_input: ValidInput = ''):
        self._most_recent_input = most_recent_input
        self._pos = pos
        self._previous_input = previous_input
        previous = get_input_as_string(previous_input)
        message = '\n'.join([
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            '',
            *build_error_position(
                f'{previous}{get_input_as_string(most_recent_input)}',
                pos + len(previous),
            ),
        ])
        super().__init__(message)
        self.name = 'InputParseError'
        self.message = message

    @property
    def src(self) -> str:
        return (
            f'{get_input_as_string(self._previous_input)}'
            f'{get_input_as_string(self._most_recent_input)}'
        )

    @property
    def error_for_most_recent_input(self) -> str:
        return '\n'.join([
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            '',
            *build_error_position(get_input_as_string(self._most_recent_input), self._pos),
        ])
